File: vitral/src/vsdk/toolkit/environment/geometry/ComputationalGeometry.java
//===========================================================================
//=-------------------------------------------------------------------------=
//= Module history:                                                         =
//= - May 10 2007 - Oscar Chavarro: Original base version                   =
//===========================================================================
package vsdk.toolkit.environment.geometry;

import vsdk.toolkit.common.VSDK;
import vsdk.toolkit.common.Vector3D;
import vsdk.toolkit.environment.geometry.Geometry;
import vsdk.toolkit.environment.geometry.InfinitePlane;

/**
This class contains common computational geometry operations.
*/
public class ComputationalGeometry
{
    private static InfinitePlane workPlane;

    static {
        workPlane = new InfinitePlane(1, 1, 1, 0);
    }

    public static int triangleContainmentTest(
        Vector3D p0, Vector3D p1, Vector3D p2, Vector3D p,
        double distanceTolerance)
    {
    Vector3D n, a, b;

    a = p1.substract(p0);
    b = p2.substract(p0);
    n = a.crossProduct(b);
    n.normalize();

        workPlane.setFromPointNormal(p0, n);

        if ( workPlane.doContainmentTest(p, distanceTolerance) == 
             Geometry.LIMIT ) {
            // Barycentric coordinates test containment technique
            Vector3D c = p.substract(p0);
        double dot00, dot01, dot02, dot11, dot12, invDenom, u, v;
        dot00 = a.dotProduct(a);
        dot01 = a.dotProduct(b);
        dot02 = a.dotProduct(c);
        dot11 = b.dotProduct(b);
        dot12 = b.dotProduct(c);
            invDenom = 1 / (dot00 * dot11 - dot01 * dot01);
            u = (dot11 * dot02 - dot01 * dot12) * invDenom;
            v = (dot00 * dot12 - dot01 * dot02) * invDenom;
            if ( (u > 0) && (v > 0) && (u + v < 1) ) {
                return Geometry.LIMIT;
        }
        }
        return Geometry.OUTSIDE;
    }
}

//===========================================================================
//= EOF                                                                     =
//===========================================================================

File: cpp/base/src/main/vsdk/toolkit/processing/ComputationalGeometry.cpp
#include "vsdk/toolkit/processing/ComputationalGeometry.h"

#include <cmath>

#include "vsdk/toolkit/common/VSDK.h"
#include "vsdk/toolkit/processing/Containment.h"

double ComputationalGeometry::lineToPointDistance(
    const Vector3Dd& p0,
    const Vector3Dd& p1,
    const Vector3Dd& p)
{
    Vector3Dd lineVector = p1.subtract(p0);
    double denominator = lineVector.length();
    if ( denominator < VSDK::EPSILON ) {
        return NAN;
    }

    Vector3Dd v = p1.subtract(p0);
    Vector3Dd w = p.subtract(p0);
    double c1 = w.dotProduct(v);
    if ( c1 <= 0.0 ) {
        return p.subtract(p0).length();
    }

    double c2 = v.dotProduct(v);
    if ( c2 <= c1 ) {
        return p.subtract(p1).length();
    }

    double b = c1 / c2;
    Vector3Dd pb = p0.add(v.multiply(b));
    return p.subtract(pb).length();
}

int ComputationalGeometry::lineContainmentTest(
    const Vector3Dd& p0,
    const Vector3Dd& p1,
    const Vector3Dd& p,
    double distanceTolerance)
{
    double d = lineToPointDistance(p0, p1, p);
    if ( d <= distanceTolerance ) {
        return static_cast<int>(Containment::LIMIT);
    }
    return static_cast<int>(Containment::OUTSIDE);
}

int ComputationalGeometry::lineSegmentContainmentTest(
    const Vector3Dd& p0,
    const Vector3Dd& p1,
    const Vector3Dd& p,
    double distanceTolerance)
{
    Vector3Dd a = p1.subtract(p0);
    Vector3Dd b = p.subtract(p0);
    double denominator = a.length();
    if ( denominator < VSDK::EPSILON ) {
        return static_cast<int>(Containment::OUTSIDE);
    }

    double numerator = a.crossProduct(b).length();
    double d = numerator / denominator;
    if ( d <= distanceTolerance ) {
        double t = a.dotProduct(b) / a.dotProduct(a);
        if ( t < -VSDK::EPSILON || t > 1 + VSDK::EPSILON ) {
            return static_cast<int>(Containment::OUTSIDE);
        }
        return static_cast<int>(Containment::LIMIT);
    }
    return static_cast<int>(Containment::OUTSIDE);
}
